## 6-е домашнее задание - Библиотека numpy

<b>Срок до 2 июня 2022г. (включительно)</b>

Подпишите этот блокнот: укажите ФИО и номер группы. Выполненное задание загрузите в свой репозиторий «<b>Spring_Python_2022</b>» в github в папку "<b>homework-6</b>".  

Перед выполнением этого домашнего задания полезно прочитать про возможности и основные методы библиотеки numpy:  
http://acm.mipt.ru/twiki/bin/view/Cintro/PythonNumpy  


In [1]:
import numpy as np

1.	Создайте равномерное разбиение интервала от -1.3 до 2.5 на 64 отрезка.

In [2]:
a = np.linspace(-1.3, 2.5, 65)
a

array([-1.3     , -1.240625, -1.18125 , -1.121875, -1.0625  , -1.003125,
       -0.94375 , -0.884375, -0.825   , -0.765625, -0.70625 , -0.646875,
       -0.5875  , -0.528125, -0.46875 , -0.409375, -0.35    , -0.290625,
       -0.23125 , -0.171875, -0.1125  , -0.053125,  0.00625 ,  0.065625,
        0.125   ,  0.184375,  0.24375 ,  0.303125,  0.3625  ,  0.421875,
        0.48125 ,  0.540625,  0.6     ,  0.659375,  0.71875 ,  0.778125,
        0.8375  ,  0.896875,  0.95625 ,  1.015625,  1.075   ,  1.134375,
        1.19375 ,  1.253125,  1.3125  ,  1.371875,  1.43125 ,  1.490625,
        1.55    ,  1.609375,  1.66875 ,  1.728125,  1.7875  ,  1.846875,
        1.90625 ,  1.965625,  2.025   ,  2.084375,  2.14375 ,  2.203125,
        2.2625  ,  2.321875,  2.38125 ,  2.440625,  2.5     ])

2.	Сгенерируйте numpy массив длины $3n$, заполненный циклически числами 1, 2, 3, 1, 2, 3, 1....

In [3]:
n = 4
b = np.tile(np.arange(1, 4), (1, n))
b

array([[1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3]])

3.	Создайте массив первых 10 нечетных целых чисел.

In [29]:
n = 10
c = np.linspace(1, 2 * n - 1, n)   
c

array([ 1.,  3.,  5.,  7.,  9., 11., 13., 15., 17., 19.])

4.  Создайте массив нулей размера 10 x 10, а затем создайте в нём "рамку" из единиц по краям.

In [8]:
# 1 вариант
d = np.zeros((10, 10))
d[:, 0:10:9] = 1
d[0:10:9, :] = 1
print('Рамка в массиве:', d, sep='\n', end='\n\n')

# 2 вариант
e = np.zeros((10, 10))
e = np.pad(e, pad_width=1, constant_values=1)
print('Рамка вокруг массива:', e, sep='\n')

Рамка в массиве:
[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]

Рамка вокруг массива:
[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]


5.  Создайте массив 8 x 8 с шахматной доской из нулей и единиц.

In [10]:
# 1 вариант
f = np.zeros((8, 8), dtype=int)
f[1::2,::2] = 1
f[::2,1::2] = 1
print(f, end='\n\n')

# 2 вариант
g = np.tile(np.array([[0,1],[1,0]]), (4,4))
print(g)

[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]

[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]


6.  Создайте функцию, которая создает $n×n$ матрицу с $(i,j)$-элементом, равным $i+j$.

In [11]:
def create_m(n):
    matrix = np.zeros((n, n))
    for i in range(n):
        matrix[i, :] = np.arange(0, n) + i
    return matrix

print(create_m(10))

[[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9.]
 [ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]
 [ 2.  3.  4.  5.  6.  7.  8.  9. 10. 11.]
 [ 3.  4.  5.  6.  7.  8.  9. 10. 11. 12.]
 [ 4.  5.  6.  7.  8.  9. 10. 11. 12. 13.]
 [ 5.  6.  7.  8.  9. 10. 11. 12. 13. 14.]
 [ 6.  7.  8.  9. 10. 11. 12. 13. 14. 15.]
 [ 7.  8.  9. 10. 11. 12. 13. 14. 15. 16.]
 [ 8.  9. 10. 11. 12. 13. 14. 15. 16. 17.]
 [ 9. 10. 11. 12. 13. 14. 15. 16. 17. 18.]]


7. Примените функцию dot для перемножения вектор на вектор, матрицу на вектор и матрицу на матрицу.

In [12]:
matrix = np.array(np.arange(0, 12)).reshape(3, 4)
print('matrix', matrix, sep='\n', end='\n\n')

vector = np.array(np.arange(0, 4))
print('vector', vector, sep='\n', end='\n\n')

print('vector.dot(vector)', vector.dot(vector), sep='\n', end='\n\n')
print('matrix.dot(vector)', matrix.dot(vector), sep='\n', end='\n\n')
print('matrix.dot(matrix.T)', matrix.dot(matrix.T), sep='\n')


matrix
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]

vector
[0 1 2 3]

vector.dot(vector)
14

matrix.dot(vector)
[14 38 62]

matrix.dot(matrix.T)
[[ 14  38  62]
 [ 38 126 214]
 [ 62 214 366]]


8. Вычислите cos(x) и sin(x) на интервале [0,1] с шагом $0.05$, а затем объедините оба массива чисел как строки в один массив.

In [13]:
cosx = np.cos(np.arange(0, 1.05, 0.05))
sinx = np.sin(np.arange(0, 1.05, 0.05))
h = np.concatenate((cosx, sinx))
h

array([1.        , 0.99875026, 0.99500417, 0.98877108, 0.98006658,
       0.96891242, 0.95533649, 0.93937271, 0.92106099, 0.9004471 ,
       0.87758256, 0.85252452, 0.82533561, 0.7960838 , 0.76484219,
       0.73168887, 0.69670671, 0.65998315, 0.62160997, 0.58168309,
       0.54030231, 0.        , 0.04997917, 0.09983342, 0.14943813,
       0.19866933, 0.24740396, 0.29552021, 0.34289781, 0.38941834,
       0.43496553, 0.47942554, 0.52268723, 0.56464247, 0.60518641,
       0.64421769, 0.68163876, 0.71735609, 0.75128041, 0.78332691,
       0.8134155 , 0.84147098])

9. Создайте матрицу $A$ размера $3×5$ из случайных чисел с равномерным распределением на отрезке $[-1,3]$ (используйте np.random.rand)

In [33]:
A = (3 - (-1)) * np.random.rand(3, 5) + (-1)
A

array([[ 0.20045632,  2.00159569, -0.20171148,  1.00747005,  0.29265013],
       [-0.82744408,  0.34198092,  0.72746583,  0.79391504,  0.35366134],
       [-0.27083771,  1.30934479, -0.78299625,  0.11184021,  2.2249511 ]])

10. Найдите сумму всех элементов, сумму внутри строк, сумму внутри столбцов, а также среднее значение, дисперсию и стандартное отклонение чисел для каждой строки матрицы $A$. (Подобно тому, как sorted имеет необязательный аргумент key=, многие функции Numpy имеют необязательный аргумент axis=)


In [35]:
print(f'сумма всех элементов {A.sum()}')
print(f'сумма элементов по столбцам {A.sum(axis=0)}')
print(f'сумма элементов по строкам {A.sum(axis=1)}')
print(f'среднее значение по строкам {A.mean(axis=1)}')
print(f'дисперсия по строкам {A.var(axis=1)}')
print(f'стандартное отклонение по строкам {A.std(axis=1)}')

сумма всех элементов 7.282341923605735
сумма элементов по столбцам [-0.89782546  3.65292141 -0.25724189  1.9132253   2.87126257]
сумма элементов по строкам [3.30046071 1.38957906 2.59230215]
среднее значение по строкам [0.66009214 0.27791581 0.51846043]
дисперсия по строкам [0.60185747 0.34000253 1.2039459 ]
стандартное отклонение по строкам [0.77579474 0.58309736 1.09724469]


11.  Отнимите от каждого элемента матрицы $A$ среднее по строке и поделите на стандартное отклонение строки.

In [37]:
for i in range(len(A)):
    A[i, :] = (A[i, :] - A.mean(axis=1)[i]) / A.std(axis=1)[i]
A

array([[-0.59247092,  1.72919908, -1.11086551,  0.44777038, -0.47363303],
       [-1.89566953,  0.10987035,  0.77096906,  0.88492809,  0.12990203],
       [-0.7193456 ,  0.72079124, -1.18611344, -0.37058299,  1.55525079]])

12. Отсортируйте матрицу $A$ по 3-ому столбцу, т.е. поменяйте местами строки матрицы так, чтобы 3-й столбец оказался отсортированным. Используйте для этого slicing + argsort + indexing.

In [38]:
A = np.array([[1, 2, 3],
              [-1, 10, 10000],
              [-1000, 0, 1001]])
ind = A[:, 2].argsort()
A[ind]

array([[    1,     2,     3],
       [-1000,     0,  1001],
       [   -1,    10, 10000]])

13. Посчитайте две матрицы: $B = A^TA$ и $C = AA^T$. Что вы можете о них сказать? Какого они размера? В чём их сходство и отличие?

In [18]:
A = np.array([[1, 2, 3, 4], 
              [5, 6, 7, 8], 
              [15, 12, 11, 113]])

B = A.T.dot(A)
C = A.dot(A.T)
print('B = A.T.dot(A)', B, sep='\n')
print()
print('C = A.dot(A.T)', C, sep='\n')

B = A.T.dot(A)
[[  251   212   203  1739]
 [  212   184   180  1412]
 [  203   180   179  1311]
 [ 1739  1412  1311 12849]]

C = A.dot(A.T)
[[   30    70   524]
 [   70   174  1128]
 [  524  1128 13259]]


Матрица В размера 4х4, т.к. умножалась матрица с 4 строками на матрицу с 4 столбцами.
Матрица С размера 3х3, т.к. умножалась матрица с 3 строками на матрицу с 3 столбцами.
Сходство - они обе квадратные и симметричные, суммы элементов на диагонали равны.

14. Найдите сумму диагональных элементов матриц $B$ и $C$.

In [19]:
print(f'сумма диагональных элементов для В = {np.diag(B).sum()}')
print(f'сумма диагональных элементов для C = {np.diag(C).sum()}')

сумма диагональных элементов для В = 13463
сумма диагональных элементов для C = 13463


15.	Посчитайте детерминант и ранг матриц $A$, $B$ и $C$, посмотрите на спектр (набор собственных значений) матриц $B$ и $C$, какие выводы вы можете сделать?  
(подсказка: используйте готовые функции из библиотеки np.linalg). 


In [154]:
print('A:', f'det = {np.linalg.det(B)}',
      f'rg = {np.linalg.matrix_rank(B)}', f'с.з. = {np.linalg.eig(B)[0]}', sep='\n')
print()
print('B:', f'det = {np.linalg.det(B)}', 
      f'rg = {np.linalg.matrix_rank(B)}', f'с.з. = {np.linalg.eig(B)[0]}', sep='\n')
print()
print('C:', f'det = {np.linalg.det(C)}', 
      f'rg = {np.linalg.matrix_rank(C)}', f'с.з. = {np.linalg.eig(C)[0]}', sep='\n')

A:
det = -4.281720942627827e-09
rg = 3
с.з. = [1.33764209e+04 8.56669313e+01 9.12121960e-01 1.20634162e-13]

B:
det = -4.281720942627827e-09
rg = 3
с.з. = [1.33764209e+04 8.56669313e+01 9.12121960e-01 1.20634162e-13]

C:
det = 1045215.9999999979
rg = 3
с.з. = [1.33764209e+04 9.12121960e-01 8.56669313e+01]


Совпали ранги и с.з. (только у С нет одного c.з.)

16. Численно посчитать определённый интеграл  3-мя методами:  
a)	Прямоугольников $\int\limits_a^b {f(x)dx \approx \Delta x\sum\limits_{k = 0}^N {f({x_k})} }$  
b)	Трапеций $\int\limits_a^b {f(x)dx \approx \frac{{\Delta x}}{2}\left( {f({x_0}) + 2\sum\limits_{k = 1}^{N - 1} {f({x_k})}  + f({x_N})} \right)}$  
c)	По правилу Симпсона  $\int\limits_a^b {f(x)dx \approx \frac{{\Delta x}}{3}\left( {f({x_0}) + 2\sum\limits_{k = 1}^{N/2 - 1} {f({x_{2k}})}  + 4\sum\limits_{k = 1}^{N/2} {f({x_{2k - 1}})}  + f({x_N})} \right)}$  


В реализации этих методов цикл for для суммирования использовать нельзя.  
Нужно посчитать значение интеграла с некоторым шагом интегрирования (например начать с 0.1), потом уменьшить его в десять раз и ещё раз посчитать и так до тех пор пока отличие в ответах будет в 5-м знаке после запятой, т.е. чтобы ошибка была меньше 1e-5. Получить ответы для 3-х разных методов и сделать соответствующие выводы.  

Далее необходимо сравнить полученные оценки с аналитическим решением (т.е. формульным, которое вы должны сами посчитать на листочке и вбить формулу-ответ).   

В качестве интеграла взять один из следующих с номером, сгенерированным случайным образом с помощью функции np.random.randint(10), в качестве seed для генератора случайных чисел взять номер своего студенческого билета.

<table width="80%"
       align="center">
    <tr>
        <td> 0) $\int\limits_3^4 {\frac{{{x^2} + 3}}{{x - 2}}dx}$ </td>
        <td> 1) $\int\limits_{ - 2}^{ - 1} {\frac{{x + 1}}{{{x^3} - {x^2}}}dx}$ </td>
        <td> 2) $\int\limits_1^2 {\frac{{{e^{1/{x^2}}}}}{{{x^3}}}dx}$ </td>
        <td> 3) $\int\limits_1^e {\frac{{\cos (\ln x)}}{x}dx}$ </td>
        <td> 4) $\int\limits_1^e {\frac{{dx}}{{x(1 + {{\ln }^2}x)}}}$ </td>
    </tr>                
    <tr>
        <td> 5) $\int\limits_0^{\pi /2} {{{\cos }^3}\alpha \;d\alpha }$ </td>
        <td> 6) $\int\limits_0^{1/3} {c{h^2}3x\;dx}$ </td>
        <td> 7) $\int\limits_2^3 {\frac{{dy}}{{{y^2} - 2y - 8}}dy}$ </td>
        <td> 8) $\int\limits_{3/4}^2 {\frac{{dx}}{{\sqrt {2 + 3x - 2{x^2}} }}}$ </td>
        <td> 9) $\int\limits_0^2 {\frac{{2x - 1}}{{2x + 1}}dx}$ </td>
        </tr>
</table>

In [20]:
np.random.seed(1032192924)
print(f'вариант {np.random.randint(10)}')

вариант 4


In [21]:
# прямоугольники
h = 0.000001
print('шаг', h)
X = np.arange(1, np.exp(1), h)
Y = 1 / (X * (1 + np.power(np.log(X), 2)))
print(f'прямоугольники {h * np.sum(Y)}')

шаг 1e-06
прямоугольники 0.7853986030194765


In [22]:
# трапеции
h = 0.000001
print('шаг', h)
X = np.arange(1, np.exp(1), h)
Y = 1 / (X * (1 + np.power(np.log(X), 2)))
print(f'трапеции {(h / 2) * (Y[0] + 2 * np.sum(Y[1:(len(Y) - 1)]) + Y[len(Y) - 1])}')

шаг 1e-06
трапеции 0.7853980110495595


In [23]:
# Симпсон
n = 800000
h = (np.exp(1) - 1) / (2 * n)
print('шаг', h)
X = np.linspace(1, np.exp(1), 2 * n + 1)
Y = 1 / (X * (1 + np.power(np.log(X), 2)))

s1 = np.sum(Y[2 : (2 * n - 1) : 2])  # от 2 до N-2 (до 2n-2) включительно
s2 = np.sum(Y[1 : (2 * n) : 2])  # от 1 до N-1 (до 2n-1) включительно

print("Симпсон", (h / 3) * (Y[0] + 2 * s1 + 4 * s2 + Y[2 * n]))

шаг 1.073926142786903e-06
Симпсон 0.7853981633974478


шаг 0.1

прямоугольники 0.8420537065276416

трапеции 0.7827317528632757

Симпсон 0.7853935518334139


шаг 0.01

прямоугольники 0.7898026744575757

трапеции 0.783877346025379

Симпсон 0.7853981629384734


шаг 0.001

прямоугольники 0.7859383795600955

трапеции 0.7853463906261692

Симпсон 0.7853981633974025

трапеции и Симсон уже на этом шаге отличаются в 5 знаке после запятой


шаг 0.0001

прямоугольники 0.78544230970081

трапеции 0.7853831121610418

Симпсон 0.7853981633974483
    

шаг 0.00001

прямоугольники 0.7854037467749472

трапеции 0.7853978270751071

Симпсон 0.7853981633974482


шаг 0.000001

прямоугольники 0.7853986030194765

трапеции 0.7853980110495595

Симпсон 0.7853981633974478

отличие в 5 знаке после запятой

Вывод: прямоугольники самые неточные, Симпсон самый точный

Aналитическое решение неопределенного интеграла = arctg(ln(x)) + с, подставляя пределы интегрирования, получим arctg(ln(e)) - arctg(ln(1)) = arctg(1) - arctg(0) = 0.7853981633974483 - 0 = 0.7853981633974483 rad

на шаге 0.000001:

решение прямоугольников отличается в 7 знаке, решение трапеций отличается в 7 знаке, решение Симпсона отличается в 15 знаке